src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs

```diff
@@ -63,41 +62,28 @@ public IEnumerable<ContentItemGroup> FindItemGroups(PatternSet definition)
             {
                 var groupPatterns = definition.GroupExpressions;
 
-                List<Tuple<ContentItem, Asset>> groupAssets = null;
+                List<(ContentItem Item, Asset Asset)> groupAssets = null;
+
                 foreach (var asset in _assets)
                 {
                     foreach (var groupPattern in groupPatterns)
                     {
                         var item = groupPattern.Match(asset.Path, definition.PropertyDefinitions);
                         if (item != null)
                         {
-                            if (groupAssets == null)
-                            {
-                                groupAssets = new List<Tuple<ContentItem, Asset>>(1);
-                            }
-
-                            groupAssets.Add(Tuple.Create(item, asset));
+                            groupAssets ??= new List<(ContentItem Item, Asset Asset)>(capacity: 1);
+                            groupAssets.Add((item, asset));
                         }
                     }
                 }
 
                 if (groupAssets?.Count > 0)
                 {
-                    foreach (var grouping in groupAssets.GroupBy(key => key.Item1, GroupComparer.DefaultComparer))
+                    foreach (var grouping in groupAssets.GroupBy(key => key.Item, GroupComparer.DefaultComparer))
                     {
-                        var group = new ContentItemGroup();
-
-                        foreach (var property in grouping.Key.Properties)
-                        {
-                            group.Properties.Add(property.Key, property.Value);
-                        }
-
-                        foreach (var item in FindItemsImplementation(definition, grouping.Select(match => match.Item2)))
-                        {
-                            group.Items.Add(item);
-                        }
-
-                        yield return group;
+                        yield return new ContentItemGroup(
+                            properties: new Dictionary<string, object>(grouping.Key.Properties),
```

Comment:
```suggestion
                            properties: grouping.Key.Properties,
```
How about the suggested change? If we can avoid creating a new dictionary that would be great.

Reply:
The previous code made a copy of this collection. Are you sure it's safe to share objects here? So many of the types here are mutable that it's difficult for me to reason about the safety of removing a defensive copy like this. I agree that if it's safe, however, it would help.

Reply:
I'd just keep the risk as minimal as possible, especially since we're in the last "easy to get in code" preview.

Reply:
Looking at the references to `properties`, it looks like the dictionary is only used for lookups. 

<img width="1421" alt="image" src="https://github.com/NuGet/NuGet.Client/assets/52756182/56c7f88f-83f9-4dc3-8ef5-fad67c060bb6">

May be the correct thing to do in this case is to change the type to `IReadOnlyDictionary` which can be done in follow-up PR.

Reply:
That looks encouraging. Sounds like a follow up for 17.8.

Let me know if there's anything else you'd like to see for this PR. Seems ready to me.

Reply:
Yup. Created https://github.com/NuGet/Home/issues/12657 to address the code review feedback.